drv/sidecar-front-io/src/transceivers.rs

```diff
@@ -695,38 +977,59 @@ impl Transceivers {
     /// Waits for all of the I2C busy bits to go low
     ///
     /// Returns a set of masks indicating which channels (among the ones active
-    /// in the input mask) have recorded FPGA errors.
+    /// in the input mask) have recorded FPGA errors. The meaning of the
+    /// returned `ModuleResult`:
+    /// success: the desired I2C transaction completed successfully
+    /// failure: there was an I2C error
+    /// error: an `FpgaError` occurred
     pub fn wait_and_check_i2c(
         &mut self,
-        mask: FpgaPortMasks,
-    ) -> Result<FpgaPortMasks, FpgaError> {
-        let mut out = FpgaPortMasks::default();
+        mask: LogicalPortMask,
+    ) -> ModuleResult {
+        let mut physical_failure: FpgaPortMasks = FpgaPortMasks::default();
+        let mut physical_error: FpgaPortMasks = FpgaPortMasks::default();
+        let phys_mask: FpgaPortMasks = mask.into();
 
         #[derive(AsBytes, Default, FromBytes)]
         #[repr(C)]
         struct StatusAndErr {
             busy: u16,
             err: [u8; 8],
         }
-        for fpga_index in mask.iter_fpgas() {
+        for fpga_index in phys_mask.iter_fpgas() {
             let fpga = self.fpga(fpga_index);
             // This loop should break immediately, because I2C is fast
             let status = loop {
                 // Two bytes of BUSY, followed by 8 bytes of error status
-                let status: StatusAndErr = fpga.read(Addr::QSFP_I2C_BUSY0)?;
+                let status = match fpga.read(Addr::QSFP_I2C_BUSY0) {
+                    Ok(data) => data,
+                    // If there is an FPGA communication error, mark that as an
+                    // error on all of that FPGA's ports
+                    Err(_) => {
+                        match fpga_index {
+                            FpgaController::Left => {
+                                physical_error.left = PhysicalPortMask(0xffff)
+                            }
+                            FpgaController::Right => {
+                                physical_error.right = PhysicalPortMask(0xffff)
+                            }
+                        };
```

Comment:
Optional:
```suggestion
                        *(match fpga_index {
                            FpgaController::Left => &mut physical_error.left,
                            FpgaController::Right => &mut physical_error.right,
                        }) = PhysicalPortMask(0xffff);
```

Reply:
On second thought, what about adding new functions to `FpgaPortMasks`?

```rust
impl FpgaPortMasks {
    fn get(&self, p: FpgaController) -> PhysicalPortMask {
        // ...
    }

    fn get_mut(&mut self, p: FpgaController) -> &mut PhysicalPortMask {
        // ...
    }
```

That would save a bunch of similar `match` statements.

Reply:
🧠 this is really cleaning up a number of gross spots. Great recommendation!